internal/transport/http2_client_test.go

```diff
@@ -0,0 +1,172 @@
+	if bytes.Equal(b, ClientPreface) {
+		incorrectPreface := "INCORRECT PREFACE\r\n\r\n"
+		n, err = cpl.Conn.Write([]byte(incorrectPreface))
+		return n, err
+	}
+	return cpl.Conn.Write(b)
+
+}
+
+func dialerClientPrefaceLength(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &clientPrefaceLengthConn{Conn: conn}, nil
+}
+
+func (fws *framerWriteSettingsConn) Write(b []byte) (n int, err error) {
+	if bytes.Equal(b, framerWriteSettings) {
+		return 0, errors.New("force error for Framer write setting")
+	}
+	return fws.Conn.Write(b)
+}
+
+func dialerFramerWriteSettings(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &framerWriteSettingsConn{Conn: conn}, nil
+}
+
+func (fwu *framerWindowUpdateConn) Write(b []byte) (n int, err error) {
+	if bytes.Equal(b, framerWindowUpdate) {
+		return 0, errors.New("force error for windowupdate")
+	}
+	return fwu.Conn.Write(b)
+}
+
+func dialerFramerWriteWindowUpdate(_ context.Context, addr string) (net.Conn, error) {
+	conn, err := net.Dial("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+	return &framerWindowUpdateConn{Conn: conn}, nil
+}
+
+func (s) TestNewHTTP2ClientTarget(t *testing.T) {
+	tests := []struct {
+		name     string
+		opts     ConnectOptions
+		expected string
+	}{
+		{
+			name:     "client-preface-write",
+			opts:     ConnectOptions{Dialer: dialerClientPrefaceWrite},
+			expected: "connection error: desc = \"transport: failed to write client preface: force error for preface write\"",
+		},
+		{
+			name:     "client-preface-length",
+			opts:     ConnectOptions{Dialer: dialerClientPrefaceLength},
+			expected: "connection error: desc = \"transport: preface mismatch, wrote 21 bytes; want 24\"",
+		},
+		{
+			name:     "framer-write-settings",
+			opts:     ConnectOptions{Dialer: dialerFramerWriteSettings},
+			expected: "connection error: desc = \"transport: failed to write initial settings frame: force error for Framer write setting\"",
+		},
+		{
+			name:     "framer-write-windowUpdate",
+			opts:     ConnectOptions{Dialer: dialerFramerWriteWindowUpdate, InitialConnWindowSize: 80000},
+			expected: "connection error: desc = \"transport: failed to write window update: force error for windowupdate\"",
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+
+			// Create a server.
+			lis, err := net.Listen("tcp", "localhost:0")
+			if err != nil {
+				t.Fatalf("Listen() = _, %v, want _, <nil>", err)
+			}
+			defer lis.Close()
+			ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(2*time.Second))
+			defer cancel()
+
+			_, err = NewClientTransport(ctx, context.Background(), resolver.Address{Addr: lis.Addr().String()}, test.opts, func(GoAwayReason) {})
+			if err == nil {
+				t.Errorf("got nil, want an error")
+			}
+			if err.Error() != test.expected {
```

Comment:
Error string is implementation-specific and isn't part a package's contract, therefore comparing error strings is not advisable.  That being said, this is an internal package - so we are in control here (presumably).

Reply:
done please check